python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -310,125 +329,97 @@ def _load_partition_from_path(
-                    context.asset_partition_keys, awaited_objects
-                )
-                if awaited_object is not None
-            }
+        return {
+            partition_key: awaited_object
+            for partition_key, awaited_object in zip(context.asset_partition_keys, awaited_objects)
+            if awaited_object is not None
+        }
+
+    def _load_partitions(self, context: InputContext) -> Dict[str, Any]:
+        # load multiple partitions
+        if not inspect.iscoroutinefunction(self.load_from_path):
+            return self.load_partitions(context)
+        else:
+            # load_from_path returns a coroutine, so we need to await the results
+            return self.load_partitions_async(context)
 
     def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
         # If no asset key, we are dealing with an op output which is always non-partitioned
         if not context.has_asset_key or not context.has_asset_partitions:
             path = self._get_path(context)
             return self._load_single_input(path, context)
         else:
+            # we are loading asset partitions
             asset_partition_keys = context.asset_partition_keys
             if len(asset_partition_keys) == 0:
                 return None
-            elif len(asset_partition_keys) == 1:
-                paths = self._get_paths_for_partitions(context)
-                check.invariant(len(paths) == 1, f"Expected 1 path, but got {len(paths)}")
-                path = next(iter(paths.values()))
-                backcompat_paths = self._get_multipartition_backcompat_paths(context)
-                backcompat_path = (
-                    None if not backcompat_paths else next(iter(backcompat_paths.values()))
-                )
-
-                return self._load_single_input(path, context, backcompat_path)
-            else:  # we are dealing with multiple partitions of an asset
-                type_annotation = context.dagster_type.typing_type
-                if type_annotation != Any and not is_dict_type(type_annotation):
-                    check.failed(
-                        "Loading an input that corresponds to multiple partitions, but the"
-                        " type annotation on the op input is not a dict, Dict, Mapping, or"
-                        f" Any: is '{type_annotation}'."
-                    )
-
-                return self._load_multiple_inputs(context)
+            else:
+                return self._load_partitions(context)
 
     def _handle_transition_to_partitioned_asset(self, context: OutputContext, path: "UPath"):
         # if the asset was previously non-partitioned, path will be a file, when it should
         # be a directory for the partitioned asset. Delete the file, so that it can be made
         # into a directory
+
+        # TODO: this might not be the case if the serialization format is already operating with directories
```

Comment:
can just drop this `TODO:` prefix if not planning to resolve

Reply:
Can you elaborate on what you mean? 

I'm not planning to do anything about this in this PR, I just left this comment for later

Reply:
I just meant you could drop the TODO: part of the comment 

I bias towards pushing to either tagging with an issue link if its something that should get addressed or leaving as a regular comment if it a vague idea of a future improvement with no clear intent of action. 

Just personal taste though, won't block merge on it.